hack/lib.sh

```diff
@@ -0,0 +1,163 @@
+  fi
+
+  # Grant permissions to the service account for the control plane to manage native GCP resources.
+  echo "Set up Service Account used by the Control Plane"
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/pubsub.admin
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/storage.admin
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/cloudscheduler.admin
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/logging.configWriter
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/logging.privateLogViewer
+
+}
+
+function init_pubsub_service_account() {
+  local project_id=${1}
+  local pubsub_service_account=${2}
+  echo "parameter project_id used when initiating data plane service account is'${project_id}'"
+  echo "parameter control_plane_service_account used when initiating pubsub service account is'${pubsub_service_account}'"
+  # Enable APIs.
+  gcloud services enable pubsub.googleapis.com
+
+  # Create the pubsub service account for the data plane if it doesn't exist
+  existing_pubsub_service_account=$(gcloud iam service-accounts list \
+    --filter="${SERVICE_ACCOUNT_EMAIL_KEY} ~ ^${pubsub_service_account}@")
+  if [ -z "${existing_pubsub_service_account}" ]; then
+    echo "Create PubSub Service Account '${pubsub_service_account}' neeeded for the Data Plane"
+    gcloud iam service-accounts create ${pubsub_service_account}
+  else
+    echo "PubSub Service Account needed for the Data Plane '${pubsub_service_account}' already existed"
+  fi
+
+  # Grant pubsub.editor role to the service account for the data plane to read and/or write to Pub/Sub.
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${pubsub_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/pubsub.editor
+
+}
+
+function enable_workload_identity(){
+  local project_id=${1}
+  local control_plane_service_account=${2}
+  local cluster_location_type=${3}
+
+
+  echo "parameter project_id used when enabling workload identity is'${project_id}'"
+  echo "parameter control_plane_service_account used when when enabling workload identity is'${control_plane_service_account}'"
+  local cluster_location_option
+  if [ ${cluster_location_type} == ${ZONAL_CLUSTER_LOCATION_TYPE} ]; then
+    cluster_location_option=zone
```

Comment:
why do we need `cluster_location_option `? why not use cluster_location_type directly?

Reply:
cluster_location_type is `zonal` or `regional` based on what pantheon showes. But in the gcloud command, it should be `zone` or `region`.

Reply:
I think you can tell if a cluster is zonal or regional by looking at its location format. But this can be a stretch goal. Feel free to dismiss.